### ISSUE TITLE
ExecutionError: Differentiate between client and server error.

### ISSUE THREAD

**Issue by GrangerAts** [graphql-dotnet/graphql-dotnet#1650]
## Summary

As the author of a GraphQL.Net API, it would be very helpful to be able to differentiate between an ExecutionError where the **Api-author screwed up** as opposed to when the **Api-client screwed up**. Presently, one needs to hard-code checks to exception messages found in the GraphQL.Net source code, which is a brittle approach.

### Basic example

Implement a `FieldAsync<ListGraphType<...>>` resolver and directly return your data from an `async` method; make the mistake of omitting the `await` keyword. `ExecutionStrategy.SetArrayItemNodes()` will raise an `ExecutionError` that says "User error: expected an IEnumerable list though did not find one."

### Motivation

The above error is obviously the Api-author's fault, not the Api-client's. As an Api-author, I'd like to have my Api clearly behave as if it's my fault, not the client's fault. I want to log that; I don't want to log `ExecutionErrors` that are not my fault.

I feel like there are 2 options:

#### 1. Use a non-ExecutionError
For Api-author errors, don't use `ExecutionError` at all. Instead use a built-in Exception (like `InvalidOperationException` or `ApplicationException`). This way when the Api-author screws up, they handle it just like any other "unhandled" exception that comes out of GraphQL.Net.

Searching on `ExecutionError` in GraphQL.Net, there's a number of other spots that raise "Api-author" level errors besides the above "Basic example". E.g. "Only add root types.", "Schema is not configured for mutations", "A schema is required.", "Must provide an enter or leave function.", "Unhandled document definition", etc.

#### 2. Sub-class ExecutionError
If the `ExecutionError` was instead a new subclass that could be used to globally determine "It's the Api-author's fault", then the GraphQL service has the opportunity to behave better.   

But it feels like `ExecutionError` is not supposed to be used for things that aren't problems caused by the Api-author. For example, there is already `InvalidValueException`, and that's (appropriately) used when it's an Api-client error (validation).

**Comment by sungam3r** [comment]
First, I can not find `InvalidValueError` in the current codebase. Second, I think (generally) there are only **Api-author errors** at the execution stage and only **Api-client errors** at the validation stage. It seems to me that you are trying to solve the problem by inappropriate means.

> Presently, one needs to hard-code checks to exception messages found in the GraphQL.Net source code, which is a brittle approach.

See this https://github.com/graphql/graphql-spec/issues/698

**Comment by GrangerAts** [comment]
Sorry, I meant `InvalidValueException` (fixed above). ([here](https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Execution/InvalidValueException.cs))

There are multiple places in the GraphQL.Net codebase where **Api-author errors** throw `ExecutionError` exceptions, which is the exact same exception thrown when the problem is an **Api-client error**; those were some of the examples I gave. I like the solution of switching those places away from using `ExecutionError` so that `ExecutionError` **always** means **Api-client error**.

**Comment by GrangerAts** [comment]
If there's another way to differentiate, please let me know so I can use it---instead of combing the GraphQL.Net source code for Exception.Message values to put into a bunch of "if" statements. Consider my example above. That sort of error has no business going to the client, but it's reported in the Errors collection just like any other request validation error. I know of no location/way to blindly catch `ExecutionError` and 100% know it's always "my fault" instead of "the client's fault".

**Comment by sungam3r** [comment]
I think that error codes will solve this issue (when accepted by graphql working group).

**Comment by GrangerAts** [comment]
That error-code spec is for client-side issues and will not solve the problem at all. 

E.g. why would the GraphQL working group define a code for "Must provide an enter or leave function."? ...or "Unhandled document definition"?  It's entirely custom to the implementation of the GraphQL.Net library.

**Comment by sungam3r** [comment]
So you can distinguish situations by the absence of known client codes. Or according to the availability of server codes, which we can add after the concept of codes appears in the specification.

**Comment by sungam3r** [comment]
So as a conclusion I suppose to use codes **inside** existing `ExecutionError` instead of scattering `ExecutionError` to different subclasses or especially to `InvalidOperationException` or other "outer" exceptions.

**Comment by GrangerAts** [comment]
I'm not in favor of sub-classing, nor am I in favor of doing "outer/inner" exceptions that involve `ExecutionError`. What I want is GraphQL.Net to simply throw a non-`ExecutionError` exception when it's the Api-author's fault. It's confusing to hide server-side errors in a client-side error object.

Here are some significant problems with putting server-side errors in a client-side object (the "code" approach):

1. GraphQL.Net already *appropriately* uses non-`ExecutionError` exceptions to indicate *Api-author* errors. (GraphQLTypeReference.cs, ExecutionStrategy.cs, and others use `InvalidOperationException`. GraphQLExtensions.cs and others use `ArgumentException`. HttpResponseStreamWriter.cs, DataLoaderContextExtensions.cs, ObjectGraphType.cs, and many more all use `ArgumentNullException`... the list goes on for 14 pages in a GitHub search.) 
2. To be consistent, you'd have to change ALL of those to instead throw `ExecutionError` with a new "code"; that's multiple breaking changes for everyone that uses GraphQL.Net.
3. Even if you did that, you still have .Net Core BCL functionality that throws regular exceptions. So then the official GraphQL.Net documentation is going to be something like this: *Handle exceptions like normal, but if it's an `ExecutionError`, go check the "code"; if that's null/empty, then treat it like a normal exception, otherwise, you need to let it get serialized/emitted to the client.*
4. Next, there's the issue that there isn't a GraphQL "working group" for standardization of error codes. The most I've found is [this discussion/proposal/RFC that was opened less than 2 weeks ago](https://github.com/graphql/graphql-spec/issues/708). It's presently just an idea; there's no actionable proposal to discuss, evaluate, or submit. And, based on the current GraphQL spec's revision cadence, it's going to be **multiple years** before this even makes it into the next working-draft.

From your posture, it appears that devs using the GraphQL.Net library have no choice but to write a helper method in our code that will inspect the `.Message` of each `ExecutionError` object returned in `ExecutionResult.Errors` against a hard-coded list of strings we manually pulled from the GraphQL.Net source code. If they're in the list, we change our server's response to be the same as if there had been an un-handled exception during Gql execution. Again: very brittle.  But on the bright side, when I search the code for `"new ExecutionError"`, there's only 2 pages of hits to consider. So now I'm off to write that hack.

**Comment by sungam3r** [comment]
> It's confusing to hide server-side errors in a client-side error object.

Why did you decide that `ExecutionError` describes client-side error object only?

1. IIRC all aforementioned _appropriately_ used exceptions are wraped inside `ExecutionError` later.
2. I do not think this will be breaking changes. Maybe not.
3. We can always use generic `INTERNAL_ERROR` code as default for whatever server unknown exception.
4. I understand that.

>  devs using the GraphQL.Net library have no choice but to write a helper method in our code that will inspect the .Message of each ExecutionError

Each `ExecutionError` has `Code` property. Could you check this instead of `Message`? You can inspect sources to check existing codes, for example, for `ValidationError`. By default `Code` equals to exception type name converted to CONSTANT_CASE.

**Comment by sungam3r** [comment]
@joemcbride Could you share your opinion, please? I can be shortsighted in my judgments.

**Comment by GrangerAts** [comment]
> Why did you decide that `ExecutionError` describes client-side error object only?

Because that's what's in the `ExecutionResult` (`.Errors` collection) that's returned to the client.

As for checking `ExecutionError.Code` to determine if the error is actually intended for the client vs the server, I don't know what to reliably check for. I just started browsing the 'master' repo source code (as of 2020-04-23) for GraphQL.Net and here's my raw/quick listing I've got so far (I have to pick this up another day, however; not done yet):



__So far, I think these are "our/my fault"__
```
"A schema is required." // DocumentExecutor.ValidateOptions()
"A query is required." // DocumentExecutor.ValidateOptions()
"User error: expected an IEnumerable list though did not find one." // ExecutionStrategy.SetArrayItemNodes()
"Cannot return null for non-null type." // ExecutionStrategy.ValidateNodeResult()
$"Abstract type {abstractType.Name} must resolve to an Object type at " + $"runtime for field {node.Parent.GraphType.Name}.{node.Name} " + $"with value '{result}', received 'null'." // ExecutionStrategy.ValidateNodeResult()
$"Runtime Object type \"{objectType}\" is not a possible type for \"{abstractType}\"." // ExecutionStrategy.ValidateNodeResult()
$"\"{result}\" value of type \"{result.GetType()}\" is not allowed for \"{objectType.Name}\". Either change IsTypeOf method of \"{objectType.Name}\" to accept this value or return another value from your resolver." // ExecutionStrategy.ValidateNodeResult()
$"Cannot convert '{serialized}' value to AST for type '{type.Name}'." // GraphQLExtensions.AstFromValue.Convert()
$"Unknown kind of type: {context.Source}" // __Type.__Type()
"Unknown kind of type: {0}".ToFormat(type) // __Type.KindForInstance()
"Unhandled document definition" // AST.Document.AddDefinition()
```

__So far, I think these are "their fault"__
```
"Unable to determine operation from query." // DocumentExecutor.ExecuteAsync()
```

__So far, I'm not sure who's fault this might be__
```
 var error = new ExecutionError(exceptionContext?.ErrorMessage ?? $"Error trying to resolve {node.Name}.", ex);  // ExecutionStrategy.ExecuteNodeAsync()
"Invariant requires an error message." // Invariant.Check()     (and the !valid one below)
```

**Comment by sungam3r** [comment]
$"Cannot convert '{serialized}' value to AST for type '{type.Name}'."   - this one can be both user-error and api-error.

I see that you want more information. So let's add it. Take for example "Cannot return null for non-null type." We can do something like
```c#
var error = new ExecutionError(
                            "Cannot return null for non-null type."
                            + $" Field: {parent.Name}, Type: {parent.FieldDefinition.ResolvedType}.") { Code = "SERVER_ERROR:123" };
```

We need to come up with a code naming system. And then use these codes in all places. I pointed you to spec proposal as an example of how such a proposal has already been raised at the specification level. We **can** do the same earlier at the GraphQL.NET implementation level. In the worst case, we just have to change some error codes after the release of a new spec draft. PRs are welcome.

**Comment by GrangerAts** [comment]
Instead of inventing the `{ Code = "SERVER_ERROR:123" }` paradigm (which isn't going to follow any GraphQL spec that may or may not eventually become ratified), just change it to a normal Exception. 

If people want to expose those exceptions to the world, they just set `ExecutionOptions.ThrowOnUnhandledException` to `false` and GraphQL.Net does what it does now: wrap the normal exception in the `ExecutionError` and it goes out in `ExeceptionResult.Errors`. But if they want to handle them, they set `ExecutionOptions.ThrowOnUnhandledException` to `true` and take care of it like they would in any other library. 

This "code" solution still requires people to filter the `ExeceptionResult.Errors` collection on strings buried down in a property of each object.  ...and then, it's a _Collection_, which means they have to do special code to manually remove the errors that aren't meant for public consumption. It's a very cumbersome and obfuscated approach to server-side error-handling.

**Comment by sungam3r** [comment]
> It's a very cumbersome approach to server-side error-handling.

 I was thinking that we are talking about client-side error-handling as well.

**Comment by GrangerAts** [comment]
No. I just want `ExecutionError` to always mean, "The client has to deal with this, not me".  If I want the client to have to deal with things that are my fault, I'd just set `ExecutionOptions.ThrowOnUnhandledException` to `false` (and in that case, I'd likely start caring about the `ExecutionOptions.ExposeExceptions` setting).  LOL.

**Comment by GrangerAts** [comment]
OK, so I have something to look at if/when someone has time. I expect that I've gotten a few things wrong; I don't know the GraphQL.Net code very well yet. Below is the first-draft of the method I'm intending to use. I also submitted a PR ( #1658 ) based on this, not expecting it gets merged but that it may make it easier/faster to review than looking through this method.

```
protected bool IsActuallyAServerSideError(ExecutionError ee)
{
	string msg = ee?.Message;
	if (msg is null)
		return false;

	// GraphQL\Execution\DocumentExecuter.cs
	if (msg.Equals("A schema is required.", System.StringComparison.OrdinalIgnoreCase)) // ValidateOptions()
		return true;
	if (msg.Equals("A query is required.", System.StringComparison.OrdinalIgnoreCase)) // ValidateOptions()
		return true;
	if (msg.Equals("Unable to determine operation from query.", System.StringComparison.OrdinalIgnoreCase)) // ExecuteAsync()
		return true;

	// GraphQL\Execution\ExecutionHelper.cs
	// These all appear to be request (syntax) errors.

	// GraphQL\Execution\ExecutionStrategy.cs
	if (msg.Equals("User error: expected an IEnumerable list though did not find one.", System.StringComparison.OrdinalIgnoreCase)) // SetArrayItemNodes()
		return true;
	if (msg.StartsWith("Cannot return null for non-null type.", System.StringComparison.OrdinalIgnoreCase)) // SetArrayItemNodes() AND ValidateNodeResult()
		return true;
	if (msg.Contains(" must resolve to an Object type at runtime for field ", System.StringComparison.OrdinalIgnoreCase)) // ValidateNodeResult()
		return true;
	if (msg.StartsWith("Runtime Object type ", System.StringComparison.OrdinalIgnoreCase)) // ValidateNodeResult()
		return true;
	if (msg.EndsWith(" to accept this value or return another value from your resolver.", System.StringComparison.OrdinalIgnoreCase)) // ValidateNodeResult()
		return true;

	// GraphQL\Introspection\__Type.cs
	// These appear to possibly be used for both request and response. "Unknown kind of type: ..."

	// GraphQL\Invariant.cs
	if (msg.Equals("Invariant requires an error message.", System.StringComparison.OrdinalIgnoreCase))
		return true;
	// GraphQL\GraphQLExtensions.cs
	if (msg.StartsWith("Must provide Input Type, cannot use:", System.StringComparison.OrdinalIgnoreCase)) // AstFromValue() (calls Invariant.Check())
		return true;
	// GraphQL\Validation\Rules\KnownArgumentNames.cs
	if (msg.Equals("Parent type must not be null.", System.StringComparison.OrdinalIgnoreCase)) // ValidateAsync() (calls Invariant.Check())
		return true;
	// GraphQL\Validation\ValidationContext.cs
	if (msg.Equals("Must provide a validation error.", System.StringComparison.OrdinalIgnoreCase)) // ReportError() (calls Invariant.Check())
		return true;

	// GraphQL\Language\AST\Document.cs
	if (msg.Equals("Unhandled document definition", System.StringComparison.OrdinalIgnoreCase)) // AddDefinition()
		return true;

	// GraphQL\Language\CoreToVanillaConverter.cs
	if (msg.StartsWith("Unknown default value: ", System.StringComparison.OrdinalIgnoreCase)) // VariableDefinition()
		return true;
	if (msg.StartsWith("Unmapped value type ", System.StringComparison.OrdinalIgnoreCase)) // Value()
		return true;
	if (msg.StartsWith("Unmapped type ", System.StringComparison.OrdinalIgnoreCase)) // Type()
		return true;
	if (msg.StartsWith("Unmapped operation type ", System.StringComparison.OrdinalIgnoreCase)) // ToOperationType()
		return true;

	// GraphQL\ObjectExtensions.cs
	// GetPropertyValue() appears to possibly be used for both request and response:  $"Unknown value '{value}' for enum '{fieldType.Name}'."

	// GraphQL\ResolveFieldContext\ResolveFieldContextExtensions.cs
	// TryAsyncResolve() only used in "GraphQL.Tests" project.

	// GraphQL\Types\GraphTypesLookup.cs
	if (msg.Equals("Only add root types.", System.StringComparison.OrdinalIgnoreCase)) // AddType()
		return true;
	if (msg.EndsWith("There is no way to resolve this possible type during execution.", System.StringComparison.OrdinalIgnoreCase)) // AddType() (3 locations), and ApplyTypeToReference() (2 locations)
		return true;
	if (msg.StartsWith("Must provide types for Union ", System.StringComparison.OrdinalIgnoreCase)) // AddType()
		return true;
	if (msg.StartsWith("Unable to resolve reference to type ", System.StringComparison.OrdinalIgnoreCase)) // ConvertTypeReference()
		return true;

	// GraphQL\Utilities\AstPrinter.cs
	if (msg.EndsWith("\" already exists!", System.StringComparison.OrdinalIgnoreCase)) // Field()
		return true;

	// GraphQL\Utilities\SchemaBuilder.cs
	// These all appear to be request (syntax) errors.

	// GraphQL\Validation\MatchingNodeVisitor.cs
	if (msg.Equals("Must provide an enter or leave function.", System.StringComparison.OrdinalIgnoreCase)) // MatchingNodeVisitor()
		return true;

	// GraphQL\Validation\Rules\KnownDirectives.cs
	// These all appear to be request (syntax) errors.

	return false;
}
```

**Comment by Shane32** [comment]
@GrangerAts I have another suggestion for you, which utilizes the existing codebase and does not require middleware:
1. Add a new class `MyExecutionError` that inherits from `ExecutionError`
2. Inside your graph resolvers, raise a `MyExecutionError` at any time when there is a "client side" error -- for example, when an invalid argument value is passed to a field resolver.
3. Set `ExposeExceptions` to false
4. Set `ThrowOnUnhandledException` to false
5. After the query executes (e.g. inside your middleware or MVC controller), but before the `ExecutionResult` is serialized, examine the `ExecutionResult.Errors` property.  Examine each of the errors and check if the type is `MyExecutionError`.  If not, log each error (including the stack trace or other data) to your database.
5. If there are ANY errors not of type `MyExecutionError`, return a 500 error to the caller instead of serializing the data and returning it.  Or, to mask the errors, create a new `ExecutionErrors` instance, fill it with masked copies of the errors, and set the `ExecutionResult.Errors` to the new instance.

**Comment by Shane32** [comment]
Personally I'd agree with your comments in general, namely that:
* `ExecutionError` (or a perhaps new class) should be reserved for "client" exceptions
* These "client" exceptions should always be passed to the client
* AST parsing errors should always be returned to the client
* A configurable property should configure if other exceptions are unhandled (thrown) or encapsulated
* It should be easy to reprogram the serialization of exceptions (how and which properties are serialized), including conditional serialization (so certain types of exceptions are masked)

I would suggest pulling both #1656 and #1658 (subject to review).  This would handle the above semantics quite nicely.  I would also set error codes for all errors generated internally by the graphql library.

**Comment by GrangerAts** [comment]
> 
> 
> @GrangerAts I have another suggestion for you, which utilizes the existing codebase and does not require middleware:
> 
>     1. Add a new class `MyExecutionError` that inherits from `ExecutionError`
> 
>     2. Inside your graph resolvers, raise a `MyExecutionError` at any time when there is a "client side" error -- for example, when an invalid argument value is passed to a field resolver.
> 
>     3. Set `ExposeExceptions` to false
> 
>     4. Set `ThrowOnUnhandledException` to false
> 
>     5. After the query executes (e.g. inside your middleware or MVC controller), but before the `ExecutionResult` is serialized, examine the `ExecutionResult.Errors` property.  Examine each of the errors and check if the type is `MyExecutionError`.  If not, log each error (including the stack trace or other data) to your database.
> 
>     6. If there are ANY errors not of type `MyExecutionError`, return a 500 error to the caller instead of serializing the data and returning it.  Or, to mask the errors, create a new `ExecutionErrors` instance, fill it with masked copies of the errors, and set the `ExecutionResult.Errors` to the new instance.

Thanks for the suggestion. However, I expect that it will subvert client-side syntax errors from other fields. 

For example, suppose someone sends a query asking for 3 fields where 1 field doesn't exist in the schema, 1 field supplies an invalid parameter value, and 1 that is perfectly valid. According to the Gql spec, they should get the "valid" field's result and 2 errors in the collection.  One error would have been generated by Gql.Net (the unknown field), and one generated by the resolver's code (the parameter). If I were to filter as you suggested, the client wouldn't know about the invalid field.

**Comment by Shane32** [comment]
Right; because some ExecutionErrors are errors that should be returned.

**Comment by GrangerAts** [comment]
> 
> 
> Right; because some ExecutionErrors are errors that should be returned.

I totally agree. That's why I'm hoping that in the future, people won't have to filter `ExecutionResult.Errors` in their controller/middleware. If they are OK with "server-side" errors being emitted, they use the Gql.Net defaults (unhandled exceptions get turned into `ExecutionError`, and "exception details" are not "exposed").

But if they need more fine-grained control for "server-side" errors, they either use the new `UnhandledExceptionDelegate` to replace it with an exception they're comfortable with (which is then wrapped in `ExecutionError`), or they set `ThrowOnUnhandledException = true` and completely handle it in their own catch block somewhere else.

**Comment by sungam3r** [comment]
@GrangerAts @Shane32 I carefully re-read this entire thread again. @GrangerAts I totally agree with you. Thanks to the latest PRs from @Shane32, I finally realized that the problem really exists. I see you closed #1658 so I close this issue as well in favor of linked PRs from @Shane32  in #1658. OK?

**Comment by Shane32** [comment]
For future reference: see PRs #1773, #1735, #1777, #1778, #1779, and #1769.
